### PR TITLE
Propose an RFCs Process

### PR DESCRIPTION
**Pull request by rylev**
As the windows-rs project has grown, we have moved from a model of shipping code as fast as possible to a community centered collaborative approach. With this transition it becomes useful to add a little bit of process for handling large and complex changes to the project. This PR proposes adding an RFC process that is deliberating modeled very closely on the Rust RFC process. For large and complex changes, this process will hopefully lead to better designs, less churn, and more community buy-in. I look forward to all of your feedback!

### CODE REVIEW

File: rfcs/0000-rfc-process.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- We may have to experiment with what constitutes a trivial change and what requires an RFC and how this differs for user facing changes vs internal changes.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Much like the Rust project, we may in the future want to adopt additional processes around stabilization and lightweight process for internal changes (a la the MCP process in Rust). Whatever we may wish to adopt in the future, it should be compatible with this RFC proposal.
```
Comment (riverar) [inline rfcs/0000-rfc-process.md:72]:
```suggestion
Much like the Rust project, we may in the future want to adopt additional processes around stabilization and lightweight process for internal changes (e.g. [major change proposals](https://github.com/rust-lang/rfcs/blob/master/text/2904-compiler-major-change-process.md)). Whatever we may wish to adopt in the future, it should be compatible with this RFC proposal.
```

File: rfcs/0000-rfc-process.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The RFC process require a certain amount of energy to maintain and also limits the amount of open experimentation that can be done in production releases. This ultimately means that development of windows-rs will slow some. However, this should be acceptable as the quality of features that do land will generally be much higher and the likelihood of needing to revert large changes will be much smaller.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The [RFC process](https://github.com/rust-lang/rfcs) has had proven success in the Rust community. While the Rust language RFC process is not perfect it has successfully ensured properties that we deem as desireable including stability, high-quality features, and wide consensus from many stakeholders.
```
Comment (riverar) [inline rfcs/0000-rfc-process.md:53]:
```suggestion
The [RFC process](https://github.com/rust-lang/rfcs) has had proven success in the Rust community. While the Rust language RFC process is not perfect it has successfully ensured properties that we deem as desirable including stability, high-quality features, and wide consensus from many stakeholders.
```

File: rfcs/0000-rfc-process.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+# Summary
+[summary]: #summary
+
+The "RFC" (request for comments) process is intended to provide a consistent and controlled path for new features to enter the windows-rs library ecosystem, so that all stakeholders can be confident about the direction the library is evolving in.
+
+# Motivation
+[motivation]: #motivation
+
+The `windows` crate has experienced rapid change and experimentation in the past 2 years. This experimentation has been
+invaluable, but it is is important to start solidifying design. The outcome of following the RFC process should ultimately be:
```
Comment (riverar) [inline rfcs/0000-rfc-process.md:15]:
```suggestion
invaluable, but it is important to start solidifying design. The outcome of following the RFC process should ultimately be:
```

**Comment by riverar** [comment]
Thanks for pinging me on this PR. I had a few nits inline and some more general comments/questions below.

`windows` and `windows-sys` crates inherit instability from upstream metadata. There's still a bunch of work to do there, in terms of ensuring the correct shape for various Windows APIs. And when metadata changes are made, those changes trickle down to the crate and incur sometimes significant architectural and/or signature changes.

Some questions:
* How do you envision solidifying the crate designs with the unstable upstream metadata relationship in mind?
* Are you proposing that all those changes would now go through the RFC process? Who's going to review them?
* What does stability look like for `windows` and `windows-sys` (and metadata)?
* Are there core features that must be in `windows` and `windows-sys` prior to 1.0?
* Does Microsoft have other dedicated people to put on this crate to support stabilization, servicing, and other overhead?

**Comment by rylev** [comment]
I think @kennykerr does a good job of answering the specific questions, but I will add one more general point that I tried to make clear in the RFC itself: while the RFC process can help achieve stabilization it is not meant as a stabilization mechanism itself. The RFC process ultimately reduces *churn* where churn is the introduction of changes that are then reverted, modified, or reintroduced quickly and often thereafter. Reducing *churn* means reducing the amount of change that comes simply from not taking the time to thoroughly discuss proposed changes, not considering alternatives, and not getting buy-in from stakeholders. How features are then stabilized is a separate question, albeit one that can (and should IMO) be answered through an RFC proposal. 

In other words, a lot of the questions brought up here about stability are really good questions that we need to find answers to, but they're orthogonal to an RFC process.

**Comment by riverar** [comment]
> In other words, a lot of the questions brought up here about stability are really good questions that we need to find answers to, but they're orthogonal to an RFC process.

I think the value in _reducing churn_ is diminished if the crate remains unstable (0.x) and cannot be relied upon due to breakage concerns.

File: rfcs/0000-rfc-process.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+[guide-level-explanation]: #guide-level-explanation
+
+Introducing new features or new internal architecture to the windows-rs ecosystem requires creating an RFC (request for comments) which allows others to comment on and suggest improvements to proposals for changes.
+
+An RFC should be opened when the following criteria are met:
+
+- The author has a fully formed idea of the feature and can express in detail how it will work.
+- There has been at least some discussion of the feature beforehand (i.e., the RFC should not be a surprise to contributors).
+- The feature has far reaching consequences to users or contributors of the windows-rs ecosystem, and it would be difficult to reverse. This criterion should be applied more strongly for user facing changes than for internal architecture.
+- There is potential for disagreement among contributors or users for how the change should be architected.
```
Comment (kennykerr) [inline rfcs/0000-rfc-process.md:30]:
```suggestion
- There is potential for disagreement among contributors or users for how the change should be designed.
```
```suggestion
- There is potential for disagreement among contributors or users for how the change should be architected.
```

Comment (kennykerr) [inline rfcs/0000-rfc-process.md:30]:
```suggestion
- There is potential for disagreement among contributors or users for how the change should be designed.
```
```suggestion
- There is potential for disagreement among contributors or users for how the change should be architected.
```

Comment (kennykerr) [inline rfcs/0000-rfc-process.md:30]:
```suggestion
- There is potential for disagreement among contributors or users for how the change should be designed.
```

Comment (kennykerr) [inline rfcs/0000-rfc-process.md:30]:
```suggestion
- There is potential for disagreement among contributors or users for how the change should be designed.
```
